utils/containers/type_container_generator.bzl

```diff
@@ -0,0 +1,67 @@
+def _type_container_generator_impl(ctx):
+    ctx.actions.expand_template(
+        template = ctx.file.source_template,
+        output = ctx.outputs.source,
+        substitutions = {"{TYPE}": str(ctx.attr.value_type), "{PARENT_DIRECTORY}": str(ctx.attr.parent_directory)},
+    )
+
+    ADDITIONAL_INCLUDE_PATH_ACTUAL = ""
+    if str(ctx.attr.additional_include_path) == "":
+        ADDITIONAL_INCLUDE_PATH_ACTUAL = ""
+    else:
+        ADDITIONAL_INCLUDE_PATH_ACTUAL = "#include " + "\"" + str(ctx.attr.additional_include_path) + "\""
+
+    ctx.actions.expand_template(
+        template = ctx.file.header_template,
+        output = ctx.outputs.header,
+        substitutions = {"{TYPE}": str(ctx.attr.value_type), "{ADDITIONAL_INCLUDE_PATH}": ADDITIONAL_INCLUDE_PATH_ACTUAL},
+    )
+
+_type_container_generator = rule(
+    implementation = _type_container_generator_impl,
+    attrs = {
+        "source": attr.string(mandatory = True),
+        "source_template": attr.label(mandatory = True, allow_single_file = True),
+        "header": attr.string(mandatory = True),
+        "header_template": attr.label(mandatory = True, allow_single_file = True),
+        "container_type": attr.string(mandatory = True),
+        "value_type": attr.string(mandatory = True),
+        "additional_include_path": attr.string(mandatory = False),
+        "additional_deps": attr.string(mandatory = False),
+        "parent_directory": attr.string(mandatory = False),
+    },
+    outputs = {
+        "source": "%{source}",
+        "header": "%{header}",
+    },
+)
+
+def type_container_generate(container_type, value_type, additional_include_path = "", additional_deps = "//utils/containers:dummy_dep", parent_directory = "utils/containers"):
```

Comment:
shouldn't this be `typed_container_generate`?

Reply:
yeah, or maybe generic_container_generate?